tests/testthat/test-subscript-tables.R

```diff
@@ -235,3 +235,106 @@ test_that("drop and exact recognised and used appropriately", {
+                       test.case, "[", row.idx,",]")
+   test_that(test.name,
+   {
+       z.stat.out <- attr(tbl[row.idx, ], "QStatisticsTestingInfo")[, "zstatistic"]
+       expected <- unclass(tbl)[row.idx, ]
+       expected[is.nan(expected)] <- NA
+       expect_equal(z.stat.out, expected, check.attributes = FALSE)
+   })
+
+   col.idx <- sample(ncol(tbl), 1)
+   test.name <- paste0("DS-3809: Column slices of simple tables: ",
+                       test.case, "[,", col.idx,"]")
+   test_that(test.name,
+   {
+       z.stat.out <- attr(tbl[, col.idx], "QStatisticsTestingInfo")[, "zstatistic"]
+       expected <- unclass(tbl)[, col.idx]
+       expected[is.nan(expected)] <- NA
+       expect_equal(z.stat.out, expected, check.attributes = FALSE)
+   })
+}
+
+set.seed(986)
+dim.lens.avail <- vapply(lapply(tbls, dim), length, 1L)
+test.cases <- names(tbls)[dim.lens.avail == 3L]
+for (test.case in test.cases)
+{
+    tbl <- tbls[[test.case]]
+    slice.indices <- vapply(dim(tbl), function(len) sample.int(len, 1), 1L)
+
+    test.name <- paste0("DS-3810: Single slices of 3D qTables: ",
+                        test.case, "[", slice.indices[1L],",,]")
+    test_that(test.name,
+              {
+                  z.stat.out <- attr(tbl[slice.indices[1L], , ], "QStatisticsTestingInfo")[, "zstatistic"]
+                  z.stat.out <- as.numeric(z.stat.out)
+                  expected <- unclass(tbl)[slice.indices[1L], , ]
+                  expected <- as.vector(t(expected))  # t() for row-major order in attr.
+                  expect_equal(is.na(expected), is.na(z.stat.out))
+                  expect_equal(z.stat.out, expected, check.attributes = FALSE)
+              })
+    test.name <- paste0("DS-3810: Single slices of 3D qTables: ",
+                        test.case, "[, ", slice.indices[2L],",]")
+    test_that(test.name,
+              {
+                  z.stat.out <- attr(tbl[, slice.indices[2L], ], "QStatisticsTestingInfo")[, "zstatistic"]
+                  z.stat.out <- as.numeric(z.stat.out)
+                  expected <- unclass(tbl)[, slice.indices[2L], ]
+                  expected <- as.vector(t(expected))  # t() for row-major order in attr.
+                  expected[is.nan(expected)] <- NA
+                  expect_equal(z.stat.out, expected, check.attributes = FALSE)
+              })
+    test.name <- paste0("DS-3810: Single slices of 3D qTables: ",
+                    test.case, "[,,", slice.indices[3L],"]")
+    test_that(test.name,
+              {
+                  z.stat.out <- attr(tbl[, , slice.indices[3L]], "QStatisticsTestingInfo")[, "zstatistic"]
+                  z.stat.out <- as.numeric(z.stat.out)
+                  expected <- unclass(tbl)[, , slice.indices[3L]]
+                  expected <- as.vector(t(expected))  # t() for row-major order in attr.
+                  expected[is.nan(expected)] <- NA
+                  expect_equal(z.stat.out, expected, check.attributes = FALSE)
+              })
+}
```

Comment:
Suggest adding tests for `character` references e.g. `table["Coke", "Hate"]` and perhaps the id matrix referencing. e.g. 2 column integer matrix for 2d tables, 3 column matrix for 3d arrays (assuming single stat) etc. and boolean array if the same dim as the input.